Competitive_Programming/C++/Maths & Number Theory/Binary_Exponentiation.md

```diff
@@ -0,0 +1,210 @@
+# Binary Exponentiation
+Binary exponentiation is an algorithm to find the power of any number a raise to an number n (a^n) in logarithmic time O(log n).
+
+# Algorithm
+Raising a to the power of n is expressed naively as <code>a<sup>n</sup> = a⋅a⋅…⋅a</code>. However, this approach is not practical for large a or n.
+
+The idea of binary exponentiation is, that we split the power using the binary representation of the exponent.
+
+Writing n in base 2, for example: 5<sup>13</sup> = 5<sup>(1101)<sub>2</sub></sup> = (5<sup>8</sup>).(5<sup>4</sup>).(5<sup>1</sup>)
+
+Since the number n has exactly ⌊log<sub>2</sub>n⌋+1 digits in base 2, we only need to perform <code>O(log n)</code> multiplications, if we know the powers a<sup>1</sup>, a<sup>2</sup>, a<sup>4</sup>, a<sup>8</sup>,…,a<sup>⌊logn⌋</sup>.
+
+So, a fast way to compute those is that an element in the sequence is just the square of the previous element.
+
+5<sup>1</sup> = 5
+
+5^2 = (5^1)^2 = 25
+
+5^4 = (5^2)^2 = 625
+
+5^8 = (5^4)^2 = 390625
+
+**The final complexity of this algorithm is O(logn):** we have to compute logn powers of a, and then have to do at most logn multiplications to get the final answer from them.
+
+Finally, the idea is as follows:
+
+a<sup>n</sup> = 1 &emsp;&emsp;&emsp;&emsp; if n = 0
+
+a<sup>n</sup> = (a<sup>((n-1)/2)</sup>)<sup>2</sup> * a &emsp;&emsp;&emsp;&emsp; if n is odd
+
+a<sup>n</sup> = (a<sup>(n/2)</sup>)<sup>2</sup> &emsp;&emsp;&emsp;&emsp; if n is even
+
```

Comment:
Add this lines like:

* if `n` is `odd`
    * `a<sup>n</sup> = (a<sup>((n-1)/2)</sup>)<sup>2</sup> * a`